File: m/rls/memov.r
file	memov - move memory
include rid:rider
include rid:medef
If meKrts
If Win
include rid:wimod
Else
include	<string.h>
End
End

code	me_mov - move overlapping memory

  func	me_mov
	src : * void fast			; source area
	dst : * void fast			; destinatation area
	cnt : size				; char count
	()  : * void				; past result destination
  is	res : * void				; result for moveup
If meKrts
If Win
	MoveMemory (dst, src, cnt)
Else
	memmove (dst, src, cnt)			; move it
End
	reply <*char>dst + cnt			; our reply value
Else
	reply dst if cnt eq 0			; ignore wacky counts
      if src gt dst				; source higher, move down
	while cnt ge 16U			; 16 at a time
	   (<*long>dst)[0] = (<*long>src)[0]	; 1
	   (<*long>dst)[1] = (<*long>src)[1]	; 2
	   (<*long>dst)[2] = (<*long>src)[2]	; 3
	   (<*long>dst)[3] = (<*long>src)[3]	; 4
	   <*char>dst += 16, <*char>src += 16	;
	.. cnt -=  16				; count it 
	while cnt--				; do remainder
	.. *(<*char>dst)++ = *(<*char>src)++	;
	reply dst				; end of result
      else					; source lower, move up
	<*char>src += cnt			; point past end
	<*char>dst += cnt			; point past end
	res = dst				; for the return value
	while cnt ge 16U			; 16 at a time
	   <*char>dst -= 16, <*char>src -= 16	;
	   (<*long>dst)[3] = (<*long>src)[3]	; 1
	   (<*long>dst)[2] = (<*long>src)[2]	; 2
	   (<*long>dst)[1] = (<*long>src)[1]	; 3
	   (<*long>dst)[0] = (<*long>src)[0]	; 4
	.. cnt -=  16				; count it 
	while cnt--				; do remainder
	.. *--(<*char>dst) = *--(<*char>src)	;
	reply res				; end of result
      end					;
End
  end


File: m/rls/rlb/memov.c
/* file -  memov - move memory */
#include "m:\rid\rider.h"
#include "m:\rid\medef.h"
/* code -  me_mov - move overlapping memory */
void *me_mov(
register void *src ,
register void *dst ,
size_t cnt )
{ void *res ;
#if meKrts
#include <string.h>
  memmove (dst, src, cnt);
  return ( (char *)dst + cnt);
#else 
  if( cnt == 0)return ( dst );
  if ( src > dst) {
    while ( cnt >= 16U) {
      ((long *)dst)[0] = ((long *)src)[0];
      ((long *)dst)[1] = ((long *)src)[1];
      ((long *)dst)[2] = ((long *)src)[2];
      ((long *)dst)[3] = ((long *)src)[3];
      (char *)dst += 16, (char *)src += 16;
      cnt -= 16; }
    while ( cnt--) {
      *((char *)dst)++ = *((char *)src)++; }
    return ( dst);
    } else {
    (char *)src += cnt;
    (char *)dst += cnt;
    res = dst;
    while ( cnt >= 16U) {
      (char *)dst -= 16, (char *)src -= 16;
      ((long *)dst)[3] = ((long *)src)[3];
      ((long *)dst)[2] = ((long *)src)[2];
      ((long *)dst)[1] = ((long *)src)[1];
      ((long *)dst)[0] = ((long *)src)[0];
      cnt -= 16; }
    while ( cnt--) {
      *--((char *)dst) = *--((char *)src); }
    return ( res);
  } 
#endif 
} 
